// Per business logic, element = lamp

syntax = "proto3";

package lighting.element;

import "location.proto";
import "asset.proto";

service Element {
    rpc GetElement (Request) returns (Reply);

    rpc ListElements (ListRequest) returns (stream List);

    rpc SearchElements (lighting.location.FilterByLocationRequest) returns (stream Reply);
}

message ListRequest {
    int32 limit = 1;
    int32 offset = 2;
}

message Request {
    int32 id = 1;
}

message Reply {
    // unique identifier within system
    int32 id = 1;

    // last known status of element
    ActivityStatus status = 2;

    // should be something helpful
    string description = 3;

    // details of the Asset that this element is connected to
    lighting.asset.Reply asset = 4;

    // where is this element?
    lighting.location.Location location = 15;
}

message List {
    repeated lighting.element.Reply elements = 1;
}

enum ActivityStatus {
    UNAVAILABLE = 0;
    ACTIVE = 1;
    INACTIVE = 2;
    DELETED = 15;
}
